src/server/gc_worker/rawkv_compaction_filter.rs

```diff
@@ -0,0 +1,455 @@
+use std::sync::Arc;
+use tikv_util::worker::{ScheduleError, Scheduler};
+use txn_types::Key;
+
+pub struct RawCompactionFilterFactory;
+
+impl CompactionFilterFactory for RawCompactionFilterFactory {
+    fn create_compaction_filter(
+        &self,
+        context: &CompactionFilterContext,
+    ) -> *mut DBCompactionFilter {
+        //---------------- GC context --------------
+        let gc_context_option = GC_CONTEXT.lock().unwrap();
+        let gc_context = match *gc_context_option {
+            Some(ref ctx) => ctx,
+            None => return std::ptr::null_mut(),
+        };
+
+        //---------------- GC context END --------------
+        let db = gc_context.db.clone();
+        let gc_scheduler = gc_context.gc_scheduler.clone();
+        let store_id = gc_context.store_id;
+        let region_info_provider = gc_context.region_info_provider.clone();
+
+        let current = ttl_current_ts();
+        let safe_point = gc_context.safe_point.load(Ordering::Relaxed);
+        let filter = RawCompactionFilter::new(
+            db,
+            safe_point,
+            gc_scheduler,
+            current,
+            context,
+            (store_id, region_info_provider),
+        );
+        let name = CString::new("raw_compaction_filter").unwrap();
+        unsafe { new_compaction_filter_raw(name, filter) }
+    }
+}
+
+struct RawCompactionFilter {
+    safe_point: u64,
+    engine: RocksEngine,
+    gc_scheduler: Scheduler<GcTask<RocksEngine>>,
+    current_ts: u64,
+    mvcc_key_prefix: Vec<u8>,
+    mvcc_deletions: Vec<Key>,
+    regions_provider: (u64, Arc<dyn RegionInfoProvider>),
+
+    // Some metrics about implementation detail.
+    versions: usize,
+    filtered: usize,
+    total_versions: usize,
+    total_filtered: usize,
+    mvcc_rollback_and_locks: usize,
+    orphan_versions: usize,
+    versions_hist: LocalHistogram,
+    filtered_hist: LocalHistogram,
+}
+
+thread_local! {
+    static STATS: CompactionFilterStats = CompactionFilterStats::default();
+}
+
+impl Drop for RawCompactionFilter {
+    // NOTE: it's required that `CompactionFilter` is dropped before the compaction result
+    // becomes installed into the DB instance.
+    fn drop(&mut self) {
+        self.raw_gc_mvcc_deletions();
+
+        self.engine.sync_wal().unwrap();
+
+        self.switch_key_metrics();
+        self.flush_metrics();
+    }
+}
+
+impl CompactionFilter for RawCompactionFilter {
+    fn featured_filter(
+        &mut self,
+        level: usize,
+        key: &[u8],
+        sequence: u64,
+        value: &[u8],
+        value_type: CompactionFilterValueType,
+    ) -> CompactionFilterDecision {
+        match self.do_filter(level, key, sequence, value, value_type) {
+            Ok(decision) => decision,
+            Err(e) => {
+                warn!("compaction filter meet error: {}", e);
+                GC_COMPACTION_FAILURE.with_label_values(&["filter"]).inc();
+                CompactionFilterDecision::Keep
```

Comment:
Consider mark `encountered_errors` state and always return Keep after encountering error, like how the original compaction filter for transactional kv does.